### ISSUE TITLE
Resizing the persistent volume in Azure AKS doesn't reflect the change at pod level.

### ISSUE THREAD

**Issue by LohithChanda** [kubernetes/kubernetes#68427]
Is this a BUG REPORT or FEATURE REQUEST?:

 /kind bug

**What happened:**
Currently I am using k8s version 1.11.2 on Azure AKS. I've tried resizing the persistent volume by adding `allowVolumeExpansion: true` for the storage class and editing the pvc to the desired size.

After restarting the pod the size of the pvc has changed to the desired size i.e from 2Ti -> 3Ti 
```
kubectl get pvc
mongo-0     Bound     pvc-xxxx   3Ti        RWO            managed-premium   1h
```

 but when i login to the pod and do a `df -h` the disk size still remains at 2Ti. 
```
kubetl exec -it mongo-0 bash
root@mongo-0:/# df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc        2.0T  372M  2.0T   1% /mongodb
```

/sig storage

**Comment by andyzhangx** [comment]
@LohithChanda Could you provide `kubectl describe PVC mongo-0` output?
If the PVC is already attached to a VM, resize azure disk PVC would fail, you need to delete the pod to let that azure disk unattached first, for details, refer to https://github.com/kubernetes/kubernetes/pull/64386

**Comment by LohithChanda** [comment]
kubectl describe pvc mongo-0
```
Name:          mongo-0
Namespace:     default
StorageClass:  managed-premium
Status:        Bound
Volume:        pvc-e8b9dd8c-b2cc-11e8-9535-c23d9c6cc1d2
Labels:        app=mongodb
               chart=mongodb-4.3.1
               component=primary
               release=mongo
Annotations:   pv.kubernetes.io/bind-completed=yes
               pv.kubernetes.io/bound-by-controller=yes
               volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/azure-disk
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      3Ti
Access Modes:  RWO
```
Just deleted the pod and retried it, looks the same 
`/dev/sdc        2.0T  402M  2.0T   1%  mongodb`

**Comment by andyzhangx** [comment]
I tried on my 1.11.2 cluster, it works well.
Below are the steps:
```
root@nginx-azuredisk:/# df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay          97G   47G   51G  49% /
tmpfs           3.5G     0  3.5G   0% /dev
tmpfs           3.5G     0  3.5G   0% /sys/fs/cgroup
/dev/sdc        2.0T   41M  2.0T   1% /mnt/disk
/dev/sda1        97G   47G   51G  49% /etc/hosts
shm              64M     0   64M   0% /dev/shm
tmpfs           3.5G   12K  3.5G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs           3.5G     0  3.5G   0% /sys/firmware

azureuser@k8s-master-34076307-0:~$ kubectl edit pvc pvc-azuredisk
persistentvolumeclaim/pvc-azuredisk edited
azureuser@k8s-master-34076307-0:~$ kubectl delete po nginx-azuredisk
pod "nginx-azuredisk" deleted
azureuser@k8s-master-34076307-0:~$ kubectl create -f https://raw.githubusercontent.com/andyzhangx/Demo/master/linux/azuredisk/nginx-pod-azuredisk.yaml
pod/nginx-azuredisk created
azureuser@k8s-master-34076307-0:~$ kubectl exec -it nginx-azuredisk -- bash
root@nginx-azuredisk:/# df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay          97G   47G   51G  49% /
tmpfs           3.5G     0  3.5G   0% /dev
tmpfs           3.5G     0  3.5G   0% /sys/fs/cgroup
/dev/sdc        3.0T   39M  3.0T   1% /mnt/disk
/dev/sda1        97G   47G   51G  49% /etc/hosts
shm              64M     0   64M   0% /dev/shm
tmpfs           3.5G   12K  3.5G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs           3.5G     0  3.5G   0% /sys/firmware
```

One possibility is that after deleting the pod, **wait for about two minutes**, make sure the disk is in detached state, and then recreate that pod.
You may also check in azure portal whether that disk size has been changed

**Comment by LohithChanda** [comment]
In my case, i am using a statefulset and was deleting just the pod after making changes to the pvc. So redeploying the statefulset with increased disk size seems to work. Thanks.

**Comment by LohithChanda** [comment]
I couldn't see any changes in the portal when i update the pvc. I had to update the disk size in portal first - edit the pvc accordingly and then deleting the pod made it to work. Thanks

**Comment by rvvincelli** [comment]
Hi, I might give more details later, but what we find surprising is that a pre-existing Azure disk, assigned with a proper direct claim to a pod, after being increased in size on the Azure side (no k8s involved) still shows the old size in a `df -h`.

Furthermore the internal driver shadows everything so I am not sure that `fdisk`ing and the sort is viable, and it feels wrong anyway.

What kind of information, and where, is trapped in k8s for a statically mounted disk to have a fixed size forever and ever? Perhaps I am missing something but it feels like this prevents onboarding external disks, forcing to create them as above dynamically, and breaks the transparency with the backend?

Please some advice from the hackers here :) @andyzhangx

**Comment by andyzhangx** [comment]
@rvvincelli k8s use `resize2fs` to grow volume size in linux, try:
`resize2fs /dev/sd?`

**Comment by rvvincelli** [comment]
@andyzhangx Thank you for your prompt answer :) Yes that is what I meant with `fdisk` and friends. But could you please confirm the following: if I manually extend the slice, doing something already happening in the driver on a virgin disk, would this change be permanent ie survive the pod?

Differently stated, can I just resize once and forget about it?

Thanks,

**Comment by andyzhangx** [comment]
@andyzhangx yes, resize once would always work, it would be permanent, and you could only change size of `unattached` azure disk.

**Comment by rvvincelli** [comment]
@andyzhangx thank you again, so basically this is it: https://docs.microsoft.com/nl-nl/azure/virtual-machines/linux/expand-disks
This goes in the direction of: https://github.com/kubernetes/kubernetes/issues/35941
It must be specified that static provisioning https://github.com/kubernetes/examples/tree/master/staging/volumes/azure_disk/static-provisioning/managed-disk comes at the risk of having to go through this: a user does not want to get trapped in syncing fs internals.

An attention point for Azure AKS users: direct hard disk modification from a flying pod did not work for me, you basically cannot unmount a volume device (I got a classic `Operation not permitted`), but no worries you can just spin up a VM, associate it as a data disk and do it from there; kill the VM when done or detach and unassign the data disk.  Might be dev-dependent but SYN.
Sorry for the OT Andy.

Thanks again,

**Comment by sharkymcdongles** [comment]
This doesn't work for me either in Azure 1.13. I do the resize and k8s says it works but inside azure it shows blob size is unchanged. I force the pod to detach by cordoning the node it is on and deleting the pod but the disk does not change size within Azure. This is a big bug imo since no pvcs can be resized when it works on other providers without issues.

**Comment by andyzhangx** [comment]
@sharkymcdongles on Azure, you cannot resize a disk when it's already attached to a node, you must delete the pod with that disk and let the disk be in `unattached` state, and then resize the PVC disk by `kubectl edit pvc ...`, I paste the guide about resize the azure disk PV below, just let me know if you run into any issue:
 - This feature is ony for azure managed disk, and if that disk is already attached to a running VM, disk resize will fail as following:
```
$ kubectl describe pvc pvc-azuredisk
Events:
  Type     Reason              Age               From           Message
  ----     ------              ----              ----           -------
  Warning  VolumeResizeFailed  51s (x3 over 3m)  volume_expand  Error expanding volume "default/pvc-azuredisk" of plugin kubernetes.io/azure-disk : disk.DisksClient#CreateOrUpdate: Failure responding to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code="OperationNotAllowed" Message="Cannot resize disk andy-mg1102-dynamic-pvc-d2d00dd9-6185-11e8-a6c3-000d3a0643a8 while it is attached to running VM /subscriptions/.../resourceGroups/.../providers/Microsoft.Compute/virtualMachines/k8s-agentpool-17607330-0."
```
 - Note: if run into such error issue, workaround is delete the pod or even `Deployment`/`StatefulSet`, make sure the disk is in `unattached` state and run `kubectl edit pvc ...` operation again.

**How to use this feature**
 - In the beginning, pls make sure the azure disk PVC is created by `kubernetes.io/azure-disk` storage class with `allowVolumeExpansion: true` (default is false)
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: hdd
provisioner: kubernetes.io/azure-disk
parameters:
  skuname: Standard_LRS
  kind: Managed
  cachingmode: None
allowVolumeExpansion: true
```
 - Before run `kubectl edit pvc pvc-azuredisk` operation, pls make sure this PVC is not mounted by any pod otherwise there would be resize error. Now run `kubectl edit pvc pvc-azuredisk` to change azuredisk PVC size from 6GB to 10GB
```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
...
...
  name: pvc-azuredisk
...
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 6Gi
  storageClassName: hdd
  volumeMode: Filesystem
  volumeName: pvc-d2d00dd9-6185-11e8-a6c3-000d3a0643a8
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 6Gi
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-05-27T08:14:34Z
    message: Waiting for user to (re-)start a pod to finish file system resize of
      volume on node.
    status: "True"
    type: FileSystemResizePending
  phase: Bound
```

 - After resized, run `kubectl describe pvc pvc-azuredisk` to check PVC status:
```
$ kubectl describe pvc pvc-azuredisk
Name:          pvc-azuredisk
Namespace:     default
StorageClass:  hdd
Status:        Bound
...
Capacity:      5Gi
Access Modes:  RWO
Conditions:
  Type                      Status  LastProbeTime                     LastTransitionTime                Reason  Message
  ----                      ------  -----------------                 ------------------                ------  -------
  FileSystemResizePending   True    Mon, 01 Jan 0001 00:00:00 +0000   Wed, 29 Aug 2018 02:29:52 +0000           Waiting for user to (re-)start a pod to finish file system resize of volume on node.
Events:
  Type       Reason                 Age    From                         Message
  ----       ------                 ----   ----                         -------
  Normal     ProvisioningSucceeded  3m57s  persistentvolume-controller  Successfully provisioned volume pvc-d7d250c1-ab32-11e8-bfaf-000d3a4e76db using kubernetes.io/azure-disk
Mounted By:  <none>
```

 - Create a pod mounting with this PVC, you will get
```
$ kubectl exec -it nginx-azuredisk -- bash
# df -h
Filesystem      Size  Used Avail Use% Mounted on
...
/dev/sdf        9.8G   16M  9.3G   1% /mnt/disk
...
```

Note: [volume expansion feature](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims) is beta in v1.11

**Comment by sharkymcdongles** [comment]
@andyzhangx 
I have volume expansion set to true and then when I do the resize within k8s it says:
`FileSystemResizePending   True    Mon, 01 Jan 0001 00:00:00 +0000   Sun, 06 Jan 2019 13:24:58 +0100           Waiting for user to (re-)start a pod to finish file system resize of volume on node.`

I then restart the pod and k8s says the filesize was completed successfully. However within Azure it shows as old size, and when I exec in the pod and do a df it still shows old size. I even tried to cordon the node the pod was on to force it to detach the volume and switch to a different node, but the result was the same. Size correct in k8s and incorrect in azure and within the pod.

Capacity:      100Gi
Access Modes:  RWO
VolumeMode:    Filesystem

**Comment by andyzhangx** [comment]
@sharkymcdongles what's your azure disk size after resize? has it changed to the new size? could you check on azure portal?

**Comment by sharkymcdongles** [comment]
@andyzhangx When I check in the Azure portal AFTER the resize says it completed successfully in kubernetes. The Azure portal still shows the OLD size. Kubernetes describe pvc shows NEW size. Container itself shows OLD size. When I try to have the pod move between nodes and the volumes as well to force a complete detach since azure disk has RWO in our cluster, it still  shows OLD size in Azure portal. So clearly something isn't working properly there.

**Comment by andyzhangx** [comment]
@sharkymcdongles thanks for the info. It's obvious that the disk size is not changed. You need to delete the pod or even `Deployment`/`StatefulSet`, make sure the disk is in `unattached` state and then run `kubectl edit pvc ...` operation again. After that operation, you could check in azure portal that the disk size has been changed to the new size. That's the correct way since azure disk could not be changed when it's already attached to a VM.

**Comment by sharkymcdongles** [comment]
@andyzhangx The way kubernetes is designed is so you can resize the PVC the restart the pod and it is now resized. Your implementation should work as follows to be of any use:

Resize issued inside kubernetes.

Resize set to pending within azure.

Pod deleted causing disk to detach.

Disk is resized in Azure.

Pod starts as expected with resized disk.

Having the completely delete a deployment is a no go for most users especially if they use an operator.

**Comment by pydevops** [comment]
@andyzhangx I also run into the same issue when I am working on a operator for influxdb. I would be nice that azure disk team can solve the issue so that we can maintain consistency. In my case, `Statefulset` is created dynamically in operator code so that I can't delete sts without losing the entire pv. So a hacky way I came up with is to make the new pod with a non existing PVC, so the new pod can be put into "pending" state and az managed disk can be resized at that time .Thanks for all the info here.

**Comment by andyzhangx** [comment]
@pydevops thanks for sharing, this could be the only workaround now.

**Comment by tumluliu** [comment]
very helpful thread, thanks a lot @andyzhangx !
A follow-up question: when I have seen the `Waiting for user to (re-)start a pod to finish file system resize of volume on node.` message, shall I just keep waiting for itself to finish the resizing, or I should launch a pod and run `resize2fs` explicitly?
The background is I am resizing a pvc from 256Gi to 1024Gi. It's been already running for a while (around 15min) and still shows me the `Waiting for user to ...` message. I have seen the disk (unattached already) size from Azure portal changed to 1TB, and the pv size is also changed to 1TB. I am just not sure if the k8s and I are waiting for each other.

**Comment by pydevops** [comment]
> very helpful thread, thanks a lot @andyzhangx !
> A follow-up question: when I have seen the `Waiting for user to (re-)start a pod to finish file system resize of volume on node.` message, shall I just keep waiting for itself to finish the resizing, or I should launch a pod and run `resize2fs` explicitly?
> The background is I am resizing a pvc from 256Gi to 1024Gi. It's been already running for a while (around 15min) and still shows me the `Waiting for user to ...` message. I have seen the disk (unattached already) size from Azure portal changed to 1TB, and the pv size is also changed to 1TB. I am just not sure if the k8s and I are waiting for each other.

I believe you need to do a `kubectl delete <pod-uses-the-pvc>`. And I don't think you need to run `resize2fs` as the new pod will mount with the new size. I could be wrong.

**Comment by tumluliu** [comment]
@pydevops I have already deleted all the pods using the pvc before I started the `kubectl edit pvc...`. And that's why the disks have been detached from the VM. And what do you mean by "the new pod"? So k8s is waiting for me to launch a pod and mount the pvc, rt?

**Comment by tumluliu** [comment]
and from this message:
```
Type       Reason                 Age    From                         Message
  ----       ------                 ----   ----                         -------
  Normal     ProvisioningSucceeded  3m57s  persistentvolume-controller  Successfully provisioned volume pvc-d7d250c1-ab32-11e8-bfaf-000d3a4e76db using kubernetes.io/azure-disk
```
I would assume the `persistentvolume-controller` was created temporarily to do the disk resizing by k8s itself? But I can't see any such temp pod in my k8s cluster now...

**Comment by pydevops** [comment]
> @pydevops I have already deleted all the pods using the pvc before I started the `kubectl edit pvc...`. And that's why the disks have been detached from the VM. And what do you mean by "the new pod"? So k8s is waiting for me to launch a pod and mount the pvc, rt?

yes correct.

**Comment by tumluliu** [comment]
thank you @pydevops . I will try that

**Comment by tumluliu** [comment]
It's just working as you mentioned 👍 . Thank you again @pydevops

**Comment by arianitu** [comment]
Hello, is there a way to get the pod to wait until the disk resize happens? Right now, after I kill the pod, it re-launches and the pvc never has time to resize.

This is troublesome because I am trying to resize a statefulset one by one. The StatefulSet is a Kafka cluser and it requires a certain amount of brokers to be online. I can't just shutdown the entire thing to resize the nodes.

I also cannot add new nodes with a bigger size since there is no way to change the volumeClaimTemplate in a statefulset.

If I reduce the number of replicas by 1, I can resize the last instance, but I can't resize the 2nd last instance. I would essentially need to shutdown half the cluster to be able to resize a disk which isn't possible.

Is there a way to prevent a pod from running / mounting so I can resize one by one?

**Comment by deepaksood619** [comment]
@andyzhangx  is resizing the disk on the fly will be available in AKS, any updates on this?

**Comment by markhley** [comment]
I had same issue with Kafka deployment on AKS. I did find a hacky way to get the PVCs resized. For each PVC, I resized the request. As stated, the Kafka PODs were bound the the respective PVCs so Azure would not increase the disk size. I placed a NoSchedule taint on every node in the cluster, thus preventing any new PODs to schedule. I then deleted all Kafka PODs. They did not restart due to taint which allowed AKS/Azure to up the size request. I then took off taint to all nodes and all Kafka PODs started and extended the file system on the resized disk. Just took a few minutes for Kafka to become healthy again. Hope this helps in similar situations.
